src/proxy/util.rs

```diff
@@ -22,3 +25,89 @@ pub fn is_runtime_shutdown(e: &Error) -> bool {
     }
     false
 }
+
+#[derive(Default)]
+pub struct Baggage {
+    pub cluster_id: Option<String>,
+    pub namespace: Option<String>,
+    pub workload_name: Option<String>,
+    pub service_name: Option<String>,
+    pub revision: Option<String>,
+}
+
+pub fn parse_baggage_header(h: Option<&HeaderValue>) -> Baggage {
+    let mut baggage = Baggage {
+        ..Default::default()
+    };
+    if let Some(hv) = h {
+        match hv.to_str() {
+            Ok(v) => {
+                v.split(&[',']).for_each(|s| {
+                    let parts: Vec<&str> = s.split('=').collect();
+                    if parts.len() > 1 {
+                        let val = match parts[1] {
+                            "" => None,
+                            s => Some(s.to_string()),
+                        };
+                        match parts[0] {
+                            "k8s.cluster.name" => baggage.cluster_id = val,
+                            "k8s.namespace.name" => baggage.namespace = val,
+                            "k8s.deployment.name"
+                            | "k8s.cronjob.name"
+                            | "k8s.pod.name"
+                            | "k8s.job.name" => baggage.workload_name = val,
+                            "service.name" => baggage.service_name = val,
+                            "service.version" => baggage.revision = val,
+                            _ => {}
+                        }
+                    }
+                });
+            }
+            Err(e) => warn!("Failed to parse baggage header: {:?}", e),
+        }
+    }
+    baggage
+}
+
+#[cfg(test)]
+pub mod tests {
+    use hyper::http::HeaderValue;
+
+    use super::parse_baggage_header;
+
+    #[test]
+    fn baggage_parser() {
+        let baggage_str = "k8s.cluster.name=K1,k8s.namespace.name=NS1,k8s.deployment.name=N1,service.name=N2,service.version=V1";
+        if let Ok(header_value) = HeaderValue::from_str(baggage_str) {
+            let baggage = parse_baggage_header(Some(&header_value));
+            assert_eq!(baggage.cluster_id, Some("K1".to_string()));
+            assert_eq!(baggage.namespace, Some("NS1".to_string()));
+            assert_eq!(baggage.workload_name, Some("N1".to_string()));
+            assert_eq!(baggage.service_name, Some("N2".to_string()));
+            assert_eq!(baggage.revision, Some("V1".to_string()));
+        };
+    }
+
+    #[test]
+    fn baggage_parser_empty_values() {
+        let baggage_str = "k8s.cluster.name=,k8s.namespace.name=,k8s.deployment.name=,service.name=,service.version=";
+        if let Ok(header_value) = HeaderValue::from_str(baggage_str) {
```

Comment:
I can't tell - do you properly handle multiline baggage header?

Reply:
Not sure too.
Are we expected to get multilined header? From the current Waypoint/Ztunnel code the header isn't multilined.
In what scenario will we get other component hbone connecting with baggage header?

Reply:
HTTP allows you to break down any header into multilines, and it's more efficient in some cases to append, that concatenate with a comma.

Reply:
I see, so the commas are replaced with new lines in such cases?
I can add a test to cover that